fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBMetaDataStore.java

```diff
@@ -723,6 +724,172 @@ public void updateRecords(@Nonnull Descriptors.FileDescriptor recordsDescriptor)
+     * Callers might need to provide an appropriate {@link MetaDataEvolutionValidator} to the meta-data store, depending on the
+     * changes that the callback performs.
+     * </p>
+     *
+     * <p>
+     * Also, callers must be cautious about modifying the stored records descriptor using mutation callbacks, as this will make the
+     * meta-data store the ultimate (and sole) source of truth for the record definitions and not just the in-database
+     * copy of a {@code .proto} file under source control.
+     * </p>
+     *
+     * @param mutateMetaDataProto a callback that mutates the meta-data proto
+     * @return a future that completes when the meta-data is mutated
+     */
+    @Nonnull
+    public CompletableFuture<Void> mutateMetaDataAsync(@Nonnull Consumer<RecordMetaDataProto.MetaData.Builder> mutateMetaDataProto) {
+        return mutateMetaDataAsync(mutateMetaDataProto, null);
+    }
+
+    /**
+     * Mutate the stored meta-data proto and record meta-data builder using mutation callbacks asynchronously.
+     *
+     * <p>
+     * This method applies the given mutation callbacks on the stored meta-data and then saves it back to the store. Note the
+     * order that the callbacks are executed. {@code mutateMetaDataProto} is called first on the meta-data proto and then the
+     * second {@code mutateRecordMetaDataBuilder} is executed on the meta-data builder.
+     * Callers might need to provide an appropriate {@link MetaDataEvolutionValidator} to the meta-data store, depending on the
+     * changes that the callbacks perform.
+     * </p>
+     *
+     * <p>
+     * Also, callers must be cautious about modifying the stored records descriptor using mutation callbacks, as this will make the
+     * meta-data store the ultimate (and sole) source of truth for the record definitions and not just the in-database
+     * copy of a {@code .proto} file under source control.
+     * </p>
+     *
+     * @param mutateMetaDataProto a callback that mutates the meta-data proto
+     * @param mutateRecordMetaDataBuilder a callback that mutates the record meta-data builder after the meta-data proto is mutated
+     * @return a future that completes when the meta-data is mutated
+     */
+    @Nonnull
+    public CompletableFuture<Void> mutateMetaDataAsync(@Nonnull Consumer<RecordMetaDataProto.MetaData.Builder> mutateMetaDataProto,
+                                                       @Nullable Consumer<RecordMetaDataBuilder> mutateRecordMetaDataBuilder) {
+        return loadCurrentProto().thenCompose(metaDataProto -> {
+            RecordMetaDataProto.MetaData.Builder metaDataBuilder = metaDataProto.toBuilder();
+            mutateMetaDataProto.accept(metaDataBuilder);
+            RecordMetaDataBuilder recordMetaDataBuilder = createMetaDataBuilder(metaDataBuilder.build());
+            recordMetaDataBuilder.setVersion(metaDataProto.getVersion() + 1);
```

Comment:
I don't think incrementing the meta-data version here is a bad idea (especially as the meta-data builder lambda is not guaranteed to increment it), but it also might make sense to increment the version within the mutation methods as well unless there's some reason to have multiple things updated together at the same meta-data version.

Reply:
Right, increasing the version here is enough for all of the methods I added. The way things stand now (pass two callbacks to the method), for `addRecordType` it's the first callback must not increment, otherwise, the second callback will set the wrong `since_version`.